### PR TITLE
Added gRPC endpoint and python client

### PR DESCRIPTION
**Pull request by johnaohara**
This PR introduces a gRPC service and an example command line client. The gRPC service uses the same hpo_service instance as the REST api, so running experiments are available simultaneously through both api's

This allows for greater integration possibilities with other tooling as the service protobuf is consumable via many different languages which allows for easy creation of HPO clients

Protobuf also handles message encoding/decoding and date type validation at the protocol level

### CODE REVIEW

**Comment by johnaohara** [comment]
I am on PTO now until 19th April, so will respond to comments when I return

**Comment by dinogun** [comment]
@johnaohara Hi John, sorry just back from PTO, I see that there is a conflict, can you please resolve that. Also we are adding a demo / example script [here](https://github.com/kruize/autotune-demo/blob/main/hpo_demo_setup.sh), would suggest that we use that repo for demo scripts. (That repo will be renamed as kruize-demos as it will be housing demos for both autotune as well as hpoaas)

**Comment by dinogun** [comment]
@chandrams can you review from a test perspective, thanks

**Comment by johnaohara** [comment]
@dinogun is the test suite working now? There are no explicit tests for this PR, but if the test-suite works, I can add them

**Comment by chandrams** [comment]
@johnaohara the testsuite is working now.

File: src/service.py [hunk 1]
```diff
@@ -14,166 +14,18 @@
...
-    return search_space_json
-
-
-
-def main():    
-    server = HTTPServer((host_name, server_port), HTTPRequestHandler)
-    logger.info("Access server at http://%s:%s" % ("localhost", server_port))
-    server.serve_forever()
+    restService.join()
+    gRPCservice.join()
```
Comment (chandrams) [inline src/service.py:28]: We will need to update all the module dependencies for gRPC in requirements.txt & in the docker file

**Comment by chandrams** [comment]
Installed these modules - grpcio, google, google-api-core

Deployed the HPO service and accessed it using the gRPC client as below

```
(base) [csubrama@csubrama hpo]$ ./deploy_hpo.sh -c native 

Using version: 0.0.1

###   Installing HPO as a native App


### Installing dependencies..........


### Starting the service...

2022-04-22 11:15:14 - INFO - rest_service - Access server at http://localhost:8085
Starting gRPC server at http://:::50051


Experiment does not exist
2022-04-22 11:15:59 - ERROR - grpc._server - Exception calling application: 'NoneType' object has no attribute 'hpo_algo_impl'
Traceback (most recent call last):
  File "/home/csubrama/anaconda3/lib/python3.8/site-packages/grpc/_server.py", line 443, in _call_behavior
    response_or_iterator = behavior(argument, context)
  File "/home/csubrama/github/john/hpo/src/grpc_service.py", line 80, in GetTrialConfig
    data = hpo_service.instance.get_trial_json_object(request.experiment_name)
  File "/home/csubrama/github/john/hpo/src/hpo_service.py", line 68, in get_trial_json_object
    if experiment.hpo_algo_impl in ("optuna_tpe", "optuna_tpe_multivariate", "optuna_skopt"):
AttributeError: 'NoneType' object has no attribute 'hpo_algo_impl'
```

```
(base) [csubrama@csubrama src]$ python grpc_client.py new --file=/home/csubrama/github/john/hpo/tests/resources/searchspace_jsons/newExperiment.json
 Adding new experiment: petclinic-sample-2-75884c5549-npvgd
Trial Number: 0
(base) [csubrama@csubrama src]$ python grpc_client.py config
 Experiment name: petclinic-sample-2-75884c5549-npvgd
 Trial number: 0
Traceback (most recent call last):
  File "grpc_client.py", line 135, in <module>
    main()
  File "/home/csubrama/anaconda3/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/csubrama/anaconda3/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/csubrama/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/csubrama/anaconda3/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/csubrama/anaconda3/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "grpc_client.py", line 87, in config
    trial_config: hpo_pb2.TrialConfig = run(fun)
  File "grpc_client.py", line 126, in run
    return func(stub)
  File "grpc_client.py", line 86, in <lambda>
    fun = lambda stub : stub.GetTrialConfig(expr)
  File "/home/csubrama/anaconda3/lib/python3.8/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/home/csubrama/anaconda3/lib/python3.8/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNKNOWN
	details = "Exception calling application: 'NoneType' object has no attribute 'hpo_algo_impl'"
	debug_error_string = "{"created":"@1650606359.811651371","description":"Error received from peer ipv4:127.0.0.1:50051","file":"src/core/lib/surface/call.cc","file_line":903,"grpc_message":"Exception calling application: 'NoneType' object has no attribute 'hpo_algo_impl'","grpc_status":2}"
```
@khansaad - John mentioned this was due to using experiment_id to lookup rather than experiment_name.  I think addressing #28 which is to replace REST APIs to use experiment_name with experiment_id will address this.

**Comment by khansaad** [comment]
Ok.
I will take that up now.

File: Dockerfile.hpo [hunk 2]
```diff
@@ -38,7 +38,7 @@ RUN microdnf install -y python3 \
 USER 1001
 
 # Install optuna to the default user
-RUN python3 -m pip install --user optuna requests scikit-optimize jsonschema
+RUN python3 -m pip install --user optuna requests scikit-optimize jsonschema click grpcio google-api-core
```
Comment (johnaohara) [inline Dockerfile.hpo:42]: @chandrams do we still need google-api-core here?

**Comment by johnaohara** [comment]
@dinogun / @chandrams are there anything outstanding issues on this PR?

**Comment by dinogun** [comment]
@johnaohara as mentioned earlier, wanted to use `experiment_name` as the key for all object retrieval before merging this. That has now been merged (PR #31). We should now be able to merge this

File: Dockerfile.hpo (hunk 2 above)
Comment (chandrams) [reply Dockerfile.hpo:42]: Without installing google-api-core when I tried I got this error:
docker run -d --name hpo_docker_container -p 8085:8085 -p 50051:50051 hpo:grpc
```
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
38a0b80b9988820a0cda7cd2930041334c3da2bc2ee0f430751673488754dd01
(base) [csubrama@csubrama hpo]$ docker logs -f hpo_docker_container
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
Traceback (most recent call last):
  File "src/service.py", line 17, in <module>
    import rest_service, grpc_service
  File "/opt/app/src/grpc_service.py", line 21, in <module>
    from gRPC import hpo_pb2, hpo_pb2_grpc
  File "/opt/app/src/gRPC/hpo_pb2.py", line 5, in <module>
    from google.protobuf import descriptor as _descriptor
ModuleNotFoundError: No module named 'google'
```

**Comment by johnaohara** [comment]
@dinogun ok, I will rebase and fix any conflicts

File: src/grpc_client.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    experiment.experiment_name = name
+    fun = lambda stub : stub.GenerateNextConfig(experiment)
+    reply: hpo_pb2.NewExperimentsReply = run(fun)
+    click.echo("Next Trial: {}".format(reply.trial_number))
+
+def run(func):
+    # NOTE(gRPC Python Team): .close() is possible on a channel and should be
+    # used in circumstances in which the with statement does not fit the needs
+    # of the code.
+    with grpc.insecure_channel('localhost:50051') as channel:
```
Comment (dinogun) [inline src/grpc_client.py:123]: Can we have the IP and PORT defined at the top and used everywhere

Comment (johnaohara) [reply src/grpc_client.py:123]: Fixed in [896cb0e](https://github.com/kruize/hpo/pull/21/commits/896cb0e12348ec9e34612760df80034e7b3b33b9) . Ideally this (and the REST port/interface) should be configurable

File: src/grpc_client.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
+# Copyright 2015 gRPC authors.
```
Comment (dinogun) [inline src/grpc_client.py:1]: Dont have a strong opinion here, but wondering if the client code should be in the scripts dir instead

**Comment by dinogun** [comment]
> > @johnaohara as mentioned earlier, wanted to use experiment_name as the key for all object retrieval before merging this. That has now been merged (PR #31). We should now be able to merge this
> > @dinogun is the plan to drop experiment_id completely?

From the input API yes, as id is something that ideally should be generated by HPOaaS. The generated id in this case is a combination of trial_id and trial_num.

File: src/grpc_client.py (hunk 4 above)
Comment (johnaohara) [reply src/grpc_client.py:1]: I did think about this, the client and server do depend on the generated gRPC python stub classes. If this was a Java project I would extract the common classes into a separate module and the client & server would depend on that module.  I am not familiar with Python best practice. Any guidance here would be appreciated.

File: Dockerfile.hpo (hunk 2 above)
Comment (johnaohara) [reply Dockerfile.hpo:42]: we don't need to pull in the full google-api (i.e to interact with with google services), we just need `protobuf` package (which is a dependency of google-api-core. I'll update the dockerfile

Comment (johnaohara) [reply Dockerfile.hpo:42]: Updated

**Comment by chandrams** [comment]
Tested it manually changes look fine.

File: src/grpc_client.py (hunk 4 above)
Comment (dinogun) [reply src/grpc_client.py:1]: ok we can take this up in a separate PR

**Review by chandrams — APPROVED** [review]
LGTM

File: src/grpc_client.py (hunk 4 above)
Comment (johnaohara) [reply src/grpc_client.py:1]: I opened an issue to track this: https://github.com/kruize/hpo/issues/37

File: src/service.py (hunk 1 above)
Comment (johnaohara) [reply src/service.py:28]: This has been resolved in Dockerfile and requirements.txt

**Review by dinogun — APPROVED** [review]
LGTM

**Comment by johnaohara** [comment]
Thanks a lot

**Comment by dinogun** [comment]
Thank you for helping HPO get off the ground!

**Pull request merged**